ProcessLib/ChemistryLib/PhreeqcInterface.cpp

```diff
@@ -0,0 +1,299 @@
+            {
+                switch (status)
+                {
+                    case Status::SettingSolutions:
+                    {
+                        // Set pH value by hydrogen concentration.
+                        auto pH =
+                            -log10((*process_solution_iter)->get(node_id));
+                        solution.setpH(pH);
+                        break;
+                    }
+                    case Status::UpdatingProcessSolutions:
+                    {
+                        // Update hydrogen concentration by pH value.
+                        auto hydrogen_concentration =
+                            pow(10, -solution.getpH());
+                        (*process_solution_iter)
+                            ->set(node_id, hydrogen_concentration);
+                        break;
+                    }
+                }
+            }
+
+            ++process_id;
+        }
+    }
+}
+
+void phreeqcIO::writeInputsToFile()
+{
+    DBUG("Writing phreeqc inputs into file '%s'.", _input_file);
+    std::ofstream outFile(_input_file, std::ofstream::out);
+    outFile << *this;
+    outFile.close();
+}
+
+std::ofstream& operator<<(std::ofstream& outFile, phreeqcIO const& phreeqcIO)
+{
+    int const nNodes = phreeqcIO._solutions.size();
+    for (int node_id = 0; node_id < nNodes; ++node_id)
+    {
+        auto const& solution = phreeqcIO._solutions[node_id];
+        outFile << "SOLUTION " << node_id + 1 << std::endl;
+        outFile << solution << std::endl;
+
+        auto const& nodal_equilibrium_phases =
+            phreeqcIO._equilibrium_phases[node_id];
+        if (!nodal_equilibrium_phases.empty())
+        {
+            outFile << "EQUILIBRIUM_PHASES " << node_id + 1 << std::endl;
+            outFile << nodal_equilibrium_phases << std::endl;
+        }
+
+        auto const& nodal_kinetic_reactant =
+            phreeqcIO._kinetic_reactants[node_id];
+        if (!nodal_kinetic_reactant.empty())
+        {
+            outFile << "KINETICS " << node_id + 1 << std::endl;
+            outFile << nodal_kinetic_reactant;
+            outFile << "-steps " << phreeqcIO._dt << std::endl << std::endl;
+        }
+
+        if (node_id == 0)
+        {
+            auto const& reaction_rates = phreeqcIO._reaction_rates;
+            if (!reaction_rates.empty())
+            {
+                outFile << "RATES" << std::endl;
+                outFile << reaction_rates << std::endl;
+            }
+
+            outFile << "SELECTED_OUTPUT" << std::endl;
+            outFile << *phreeqcIO._output;
+        }
+
+        outFile << "END" << std::endl << std::endl;
+    }
+
+    return outFile;
+}
+
+void phreeqcIO::execute()
+{
+    DBUG("Phreeqc: Executing chemical calculation.");
+    // initialize phreeqc configurations
+    auto const instance_id = CreateIPhreeqc();
+
+    // load a specific database in the working directory
+    (void)LoadDatabase(instance_id, _database.c_str());
```

Comment:
Why void cast?

Reply:
Dropped casting on the return type and introduced a check on the loading state of specified database.